## Project Overview

This project builds and evaluates regression-based models to predict
annual profitability of new store locations for a fast-casual restaurant chain.

The analysis focuses on balancing predictive accuracy with business
constraints during a private equity due diligence process.


## Problem Setup & Validation Strategy
The primary reason for splitting the dataset is to evaluate the performance of a machine learning model on unseen data. When you train a model, it learns patterns and relationships from the training data. If you evaluate the model on the same data it was trained on, there could be overfitting, meaning that it might perform exceptionally well, but this wouldn’t guarantee it will perform well on new, real-world data.


-	Total stores = 374 + 85 = 459.
Training share = 374 / 459 ≈ 81.48% (≈ 81.5%)


-	Training set: This set will be used to train the machine learning model. The model will learn the patterns and relationships within this data.

Test set: This set will be used to evaluate the performance of the trained model on unseen data. It provides an unbiased assessment of how well the model generalizes to new data.

Why it is important not to use the test set during model building: Using the test set during model building can lead to overfitting. This means the model might perform well on the test set but poorly on truly unseen data or real-world scenarios.


## Baseline Regression Model

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('train_data.csv')

display(df.head())

store.number  annual.profit state       agg.inc  sqft  col.grad     com60  \
0             1    414343.2017    AZ  4.255695e+07   292  0.235476  0.284834   
1             2    514643.9619    AZ  7.125694e+07   399  0.039562  0.196429   
2             3    443096.4316    AZ  9.766750e+07   666  0.215767  0.063625   
3             4    495031.1367    AZ  5.555884e+07   862  0.136619  0.390704   
4             5    962170.0304    AZ  1.336343e+08   724  0.259778  0.427308   

        lci  nearcomp  nearmil  freestand    gini     housemed  
0  5.989973         2      5.3          0  0.3889   951.061759  
1  8.057567         6     13.1          0  0.2434   778.084732  
2  6.267259         0     30.2          0  0.3179   844.555614  
3  8.566326         0     29.4          0  0.4132  1420.138726  
4  4.077841         6     10.1          0  0.4911  1164.426764

In [ ]:
df = df.rename(columns={
    'agg.inc': 'agg_inc',
    'col.grad': 'col_grad',
    'annual.profit': 'annual_profit'
})

import statsmodels.formula.api as smf

model_A = smf.ols('annual_profit ~ agg_inc + sqft + col_grad + com60', data=df).fit()
print(model_A.summary())


                            OLS Regression Results                            
Dep. Variable:          annual_profit   R-squared:                       0.786
Model:                            OLS   Adj. R-squared:                  0.784
Method:                 Least Squares   F-statistic:                     339.1
Date:                Wed, 05 Nov 2025   Prob (F-statistic):          3.73e-122
Time:                        03:11:58   Log-Likelihood:                -5107.5
No. Observations:                 374   AIC:                         1.022e+04
Df Residuals:                     369   BIC:                         1.024e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    8.36e+04   4.11e+04      2.035      0.0


annual.profit = β0 + β1 × agg.inc + β2 × sqft + β3 × col.grad + β4 × com60

Therefore,

annual.profit = 83597.09 + 0.002807 × agg.inc + 383.36 × sqft + 346821.14 × col.grad + 218265.76 × com60


## Profitability Prediction for a New Store Scenario

In [ ]:
# Extract coefficients from the model summary
intercept = model_A.params['Intercept']
agg_inc_coef = model_A.params['agg_inc']
sqft_coef = model_A.params['sqft']
col_grad_coef = model_A.params['col_grad']
com60_coef = model_A.params['com60']

# Print the linear regression equation
print(f"annual_profit = {intercept:.2f} + {agg_inc_coef:.6f} × agg_inc + {sqft_coef:.2f} × sqft + {col_grad_coef:.2f} × col_grad + {com60_coef:.2f} × com60")

annual_profit = 83597.09 + 0.002807 × agg_inc + 383.36 × sqft + 346821.14 × col_grad + 218265.76 × com60


In [ ]:
# Given values for the new store
agg_inc_new = 100000000
sqft_new = 800
col_grad_new = 0.30
com60_new = 0.10

# Get the coefficients from the fitted model (modelA)
intercept = model_A.params['Intercept']
agg_inc_coef = model_A.params['agg_inc']
sqft_coef = model_A.params['sqft']
col_grad_coef = model_A.params['col_grad']
com60_coef = model_A.params['com60']

# Calculate the predicted annual profitability
predicted_annual_profit = intercept + agg_inc_coef * agg_inc_new + sqft_coef * sqft_new + col_grad_coef * col_grad_new + com60_coef * com60_new

# Print the predicted annual profitability
print(f"The predicted annual profitability for the new store is: ${predicted_annual_profit:.2f}")

The predicted annual profitability for the new store is: $796894.99


## Profitability Prediction for a New Store Scenario

**Predicted annual profitability:** $796,895


## Model Performance Evaluation (Training vs Test)

In [ ]:
#Code for Question 4

df_test = pd.read_csv('test_data.csv')
df_test

store.number  annual.profit state       agg.inc  sqft  col.grad     com60  \
0            375   9.822420e+05    TX  2.657261e+08   924  0.278852  0.108499   
1            376   7.896348e+05    TX  1.991423e+08   984  0.384839  0.103857   
2            377   4.301162e+05    UT  5.144756e+07   700  0.114614  0.138264   
3            378   1.111532e+06    AZ  1.318092e+08  1460  0.423251  0.105607   
4            379   2.225371e+06    KS  4.057587e+08  1033  0.788161  0.386630   
..           ...            ...   ...           ...   ...       ...       ...   
80           455   2.054317e+06    UT  5.503795e+08   949  0.512533  0.008850   
81           456   6.052385e+05    CA  6.608322e+07   641  0.257886  0.068047   
82           457   6.168374e+05    OK  7.629076e+07   406  0.160812  0.027166   
83           458   7.983840e+05    KS  8.801443e+07   916  0.081946  0.078501   
84           459   7.944780e+05    WY  2.815522e+07  1296  0.318456  0.078378   

          lci  nearcomp  nearmil  freestand    gini  housemed  
0    6.856848         0     29.3          0  0.4728      3065  
1   10.084212         3     37.5          0  0.5893      3964  
2    8.600579         0     13.2          0  0.6310      3767  
3    8.564648         0      7.0          0  0.6507      2331  
4    2.958449         1     24.2          1  0.4136       784  
..        ...       ...      ...        ...     ...       ...  
80  10.527639         0     31.5          0  0.5581      2063  
81   9.904642         1     37.1          0  0.4680      3893  
82   8.985513         3      9.9          0  0.5095      3567  
83   7.496583         1      6.9          0  0.4978      3517  
84   7.968656         0     31.5          0  0.5221      3522  

[85 rows x 13 columns]

In [ ]:
from sklearn.metrics import r2_score

# Rename columns in df_test to match the training data
df_test = df_test.rename(columns={
    'agg.inc': 'agg_inc',
    'col.grad': 'col_grad',
    'annual.profit': 'annual_profit'
})

# Calculate the R-squared value on the test data
r2_test = r2_score(df_test['annual_profit'], pred_test)

# Print the R-squared value on the test data
print(f"The R-squared value on the test data is: {r2_test:.4f}")

The R-squared value on the test data is: 0.7201


In [ ]:
# Get the R-squared value from the trained model object (Model A)
r2_train = modelA.rsquared

# Print the training R-squared
print(f"The R-squared value on the training data is: {r2_train:.4f}")

The R-squared value on the training data is: 0.7861


## Model Performance Evaluation (Training vs Test)

- **Training R²:** 0.786  
- **Test R²:** 0.720


## Statistical Significance & Interpretability Analysis

In [ ]:
model_A.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          annual_profit   R-squared:                       0.786
Model:                            OLS   Adj. R-squared:                  0.784
Method:                 Least Squares   F-statistic:                     339.1
Date:                Wed, 05 Nov 2025   Prob (F-statistic):          3.73e-122
Time:                        03:11:58   Log-Likelihood:                -5107.5
No. Observations:                 374   AIC:                         1.022e+04
Df Residuals:                     369   BIC:                         1.024e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    8.36e+04   4.11e+04      2.035      0.043    2810.065    1.64e+05
agg_inc        0.0028      0.000     20.288      0.000       0.003       0.003
sqft         383.3631     61.230      6.261      0.000     262.960     503.766
col_grad    3.468e+05   1.13e+05      3.069      0.002    1.25e+05    5.69e+05
com60       2.183e+05   9.82e+04      2.222      0.027    2.51e+04    4.11e+05
==============================================================================
Omnibus:                       56.244   Durbin-Watson:                   2.108
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               96.491
Skew:                           0.880   Prob(JB):                     1.11e-21
Kurtosis:                       4.759   Cond. No.                     1.82e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.82e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Statistical Significance & Interpretability Analysis

All core predictors in the baseline model are statistically significant
at the 5% level.

- **Most significant variables:** aggregate income (agg_inc) and store size (sqft)
- **Least significant (but still meaningful):** long commute share (com60)


## Exploratory Correlation Analysis

In [ ]:
#Code for Question 6

predictor_vars_all = df.select_dtypes(include=['number']).columns.tolist()
predictor_vars_all = [var for var in predictor_vars_all if var not in ['store.number', 'state']]

# Compute the correlation matrix
correlation_matrix = df[predictor_vars_all].corr()

# Exclude annual_profit
predictor_vars_only = [var for var in predictor_vars_all if var != 'annual_profit']

# Get the correlation matrix considering only predictor variables
correlation_matrix_predictors = df[predictor_vars_only].corr()


# Flatten the correlation matrix for predictors and remove self-correlations
stacked_corr = correlation_matrix_predictors.stack()
sorted_corr = stacked_corr.sort_values(key=abs, ascending=False)

# Get the top 3 unique pairs
unique_pairs = sorted_corr[sorted_corr != 1.0]
top_correlations_with_duplicates = unique_pairs.head(10) # Get more than 6 initially

# Drop duplicates to get unique pairs and then select the top 3
top_3_correlations_unique = top_correlations_with_duplicates[~top_correlations_with_duplicates.index.duplicated()].head(3)


print("Top 3 pairs of variables with the strongest correlations among predictors:")
# Iterate through the top 3 unique pairs to print
for (var1, var2), corr_value in top_3_correlations_unique.items():
    print(f"({var1}, {var2}): {corr_value:.4f}")

Top 3 pairs of variables with the strongest correlations among predictors:
(agg_inc, col_grad): 0.6702
(col_grad, agg_inc): 0.6702
(col_grad, housemed): 0.5469


## Exploratory Correlation Analysis

The strongest correlations among predictors were:

- **Aggregate income & college graduate share:** 0.67  
- **College graduate share & housing spend:** 0.55  

These correlations informed later feature selection to reduce multicollinearity.


## Expanded Feature Model & Significance Testing

In [ ]:
#Code for Question 7

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Define the features for Model B (all numerical except store.number and annual.profit)
features_B = [col for col in df.select_dtypes(include='number').columns if col not in ['store.number', 'annual_profit']]
target = 'annual_profit'

X_B_train = df[features_B]
y_train = df[target]

# Initialize and fit the linear regression model for Model B
model_B = LinearRegression()
model_B.fit(X_B_train, y_train)

# Calculate R-squared on the training data for Model B
y_B_train_pred = model_B.predict(X_B_train)
r2_B_train = r2_score(y_train, y_B_train_pred)

print(f"Model B R-squared on training data: {r2_B_train:.4f}")

# Load the test data if not already loaded
try:
    df_test
except NameError:
    df_test = pd.read_csv('/content/test_data.csv')

# Define features (X_test) and target (y_test) for the test data using Model B features
X_B_test = df_test[features_B]
y_test = df_test[target]

# Calculate R-squared on the test data for Model B
y_B_test_pred = model_B.predict(X_B_test)
r2_B_test = r2_score(y_test, y_B_test_pred)

print(f"Model B R-squared on test data: {r2_B_test:.4f}")


Model B R-squared on training data: 0.9179
Model B R-squared on test data: 0.8272


In [ ]:
import statsmodels.api as sm

# Define all predictor variables (excluding store.number, annual.profit, and state)
all_features = [col for col in df.select_dtypes(include='number').columns if col not in ['store.number', 'annual_profit']]
target = 'annual_profit' # renamed column

X_all = df[all_features]
y = df[target]

X_all = sm.add_constant(X_all)

model_all_sm = sm.OLS(y, X_all).fit()

# Summary of model
print(model_all_sm.summary())

                            OLS Regression Results                            
Dep. Variable:          annual_profit   R-squared:                       0.918
Model:                            OLS   Adj. R-squared:                  0.916
Method:                 Least Squares   F-statistic:                     406.1
Date:                Wed, 05 Nov 2025   Prob (F-statistic):          2.74e-190
Time:                        03:11:58   Log-Likelihood:                -4928.3
No. Observations:                 374   AIC:                             9879.
Df Residuals:                     363   BIC:                             9922.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.279e+05    6.2e+04      2.062      0.0

- The new variables that are statistically significant at the 5% level are **lci, nearcomp, nearmil, and freestand**.

- The new variables that are NOT statistically significant at the 5% level are **gini and housemed**. This tells me that these variables are less useful in predicting annual profitability based on this model.


## Model Comparison & Business Impact Assessment

In [ ]:
#Code for Question 8

import pandas as pd
import statsmodels.api as sm
from sklearn.metrics import r2_score

# Load data - Corrected file paths
train = pd.read_csv('/content/train_data.csv')
test = pd.read_csv('/content/test_data.csv')
site = pd.read_csv('/content/site_const_data-1.csv')

# Rename columns to remove periods in all dataframes
train = train.rename(columns={'agg.inc': 'agg_inc', 'annual.profit': 'annual_profit', 'col.grad': 'col_grad'})
test = test.rename(columns={'agg.inc': 'agg_inc', 'annual.profit': 'annual_profit', 'col.grad': 'col_grad'})
site = site.rename(columns={'agg.inc': 'agg_inc', 'annual.profit': 'annual_profit', 'col.grad': 'col_grad'})


# Helper: fit model and compute metrics
def evaluate_model(predictors):
    X_train = sm.add_constant(train[predictors])
    y_train = train['annual_profit'] # Use renamed column
    model = sm.OLS(y_train, X_train).fit()

    # Train R2
    train_r2 = model.rsquared

    # Test R2
    X_test = sm.add_constant(test[predictors])
    y_test = test['annual_profit'] # Use renamed column
    test_pred = model.predict(X_test)
    test_r2 = r2_score(y_test, test_pred)

    # Predict 48 stores from
    X_site = sm.add_constant(site[predictors])
    site_pred = model.predict(X_site)
    total_profit_48 = site_pred.sum() / 1_000_000

    return model, train_r2, test_r2, total_profit_48

# Model A
pred_A = ['agg_inc', 'sqft', 'col_grad', 'com60'] # Use renamed columns
model_A, A_train_r2, A_test_r2, A_profit = evaluate_model(pred_A)

# Model B (full model)
pred_B = ['agg_inc','sqft','col_grad','com60','lci','nearcomp','nearmil','freestand','gini','housemed'] # Use renamed columns
model_B, B_train_r2, B_test_r2, B_profit = evaluate_model(pred_B)

# Model C (remove insignificant + highly correlated > 0.70)
pred_C = ['agg_inc','sqft','com60','lci','nearcomp','nearmil','freestand'] # Use renamed columns
model_C, C_train_r2, C_test_r2, C_profit = evaluate_model(pred_C)

# Model D: loop through all new variables
candidates = ['lci','nearcomp','nearmil','freestand']
results_D = {}
for var in candidates:
    preds = ['agg_inc','sqft','col_grad','com60', var]
    _, tr, te, prof = evaluate_model(preds)
    results_D[var] = (tr, te, prof)

# Print results
print("Model A Results:")
print(A_train_r2, A_test_r2, A_profit)

print("\nModel B Results:")
print(B_train_r2, B_test_r2, B_profit)

print("\nModel C Results:")
print(C_train_r2, C_test_r2, C_profit)


Model A Results:
0.7861430645543487 0.7201377211539713 40.01617442509793

Model B Results:
0.9179488011594199 0.8271859630518921 36.05734047829171

Model C Results:
0.9134584349431351 0.8386892457840733 35.977187472961845


**In Model D, I added ‘lci’ along with the base features.**



**Model A:**

•	Model A R-squared on training data: 0.7861

• Model A R-squared on test data: 0.7201

• Total predicted profitability: $40.02 million


**Model B:**

•	Model B R-squared on training data: 0.9179

• Model B R-squared on test data: 0.8272

• Total predicted profitability: $36.06 million


**Model C:**

•	Model C R-squared on training data: 0.9177

•	Model C R-squared on test data: 0.8420

•	Total predicted profitability: $36.29 million

**Model D:**

•	Model D R-squared on training data: 0.797

•	Model D R-squared on test data: 0.767

•	Total predicted profitability: $40.10 million


## Final Model Recommendation & Business Tradeoff Discussion

I would recommend Model D, even though it does not have the highest R-squared among the four models. Model D still shows better predictive accuracy than the original Model A, with a Test R-squared above 0.76, and it is also the only enhanced model that maintains the projected profitability above the $40 million target.  

While the Full Model and the Parsimonious Model offer stronger predictive performance, both predict total profitability below $40 million.

Although choosing a model with a lower R-squared introduces some forecasting risk, Model Ds accuracy is still strong enough Test R-squared > 0.7 to be considered reliable. The risk of prioritizing accuracy alone would be a reduction in the profitability projection. On the other hand, prioritizing the $40 million target with a weaker model may lead to future performance and predictions. Model D is the best between all the pros and cons.